mmv1/products/vertexai/api.yaml

```diff
@@ -433,4 +433,170 @@ objects:
             description: |
               The disk utilization of the MetadataStore in bytes.
             output: true
+# Vertex AI Matching Engine Index
+  - !ruby/object:Api::Resource
+    name: Index
+    base_url: projects/{{project}}/locations/{{region}}/indexes
+    create_url: projects/{{project}}/locations/{{region}}/indexes
+    self_link: projects/{{project}}/locations/{{region}}/indexes/{{name}}
+    update_verb: :PATCH
+    create_verb: :POST
+    references: !ruby/object:Api::Resource::ReferenceLinks
+      api: https://cloud.google.com/vertex-ai/docs/reference/rest/v1/projects.locations.indexes/
+    async: !ruby/object:Api::OpAsync
+      operation: !ruby/object:Api::OpAsync::Operation
+        path: 'name'
+        base_url: '{{op_id}}'
+        wait_ms: 1000
+      result: !ruby/object:Api::OpAsync::Result
+        path: 'response'
+        resource_inside_response: true
+      status: !ruby/object:Api::OpAsync::Status
+        path: 'done'
+        complete: True
+        allowed:
+          - True
+          - False
+      error: !ruby/object:Api::OpAsync::Error
+        path: 'error'
+        message: 'message'
+    description: |-
+      A representation of a collection of database items organized in a way that allows for approximate nearest neighbor (a.k.a ANN) algorithms search.
+    parameters:
+      - !ruby/object:Api::Type::String
+        name: region
+        description: The region of the Metadata Store. eg us-central1
+        url_param_only: true
+        input: true
+    properties:
+      - !ruby/object:Api::Type::String
+        name: 'name'
+        description: The resource name of the Index. This value is set by Google.
+        output: true
+      - !ruby/object:Api::Type::String
+        name: 'displayName'
+        description: The display name of the Index. The name can be up to 128 characters long and can be consist of any UTF-8 characters.
+        required: true
+      - !ruby/object:Api::Type::String
+        name: 'description'
+        description: The description of the Index.
+        # Please take a look at the following links for the original definition:
+        # https://cloud.google.com/vertex-ai/docs/matching-engine/create-manage-index#create_index-drest
+        # https://cloud.google.com/vertex-ai/docs/matching-engine/configuring-indexes
+      - !ruby/object:Api::Type::NestedObject
+        name: 'metadata'
+        description: An additional information about the Index
+        properties:
+          - !ruby/object:Api::Type::String
+            name: 'contentsDeltaUri'
+            input: true
```

Comment:
I see this is writeOnly in the schema definition, but it also has a description about inserting / updating / deleting the contents of the index. `input: true` means that we'll recreate the resource if the field changes. Is that correct, or would an update via `patch` be considered okay for a `writeOnly` field like this?

I'm not super familiar with this API so I may be missing some context.

Reply:
Although I haven't tested the `path` method, I found that https://github.com/GoogleCloudPlatform/magic-modules/pull/6661 disallows Endpoint, Datasets, FeaturestoreEntitytype, FeaturestoreEntitytypeFeature to use the `patch` method. I couldn't find the details behind that on the PR. Do you happen to know why `patch` was disallowed?

Reply:
@trodge do you have any additional context here?

Reply:
I confirmed `patch` succeeded without any errors although it seems the index itself is not updated immediately based on `gcloud ai indexes list`. In addition, I found out [this official doc](https://cloud.google.com/vertex-ai/docs/matching-engine/update-rebuild-index#update_index_content_with_batch_updates) explains contentsDeltaUri is available for `patch`. I'll remove `input: true`. Thanks!

<details>
<summary>operation logs</summary>

```
➜  magic-modules git:(vertex-ai-index) ✗
gcloud ai indexes list --project=kouzoh-p-kohama --region=us-west1
Using endpoint [https://us-west1-aiplatform.googleapis.com/]
---
createTime: '2022-11-13T15:44:06.169597Z'
displayName: a test index (take 2)
etag: AMEw9yNzfYSMa5zmbeYhb_m2RoUmHLg5j9MORpVGYKiIk8r-ikQv-ePsLXcQKfwuj1G8
indexStats:
  shardsCount: 1
  vectorsCount: '2'
indexUpdateMethod: BATCH_UPDATE
metadata:
  config:
    algorithmConfig:
      treeAhConfig:
        leafNodeEmbeddingCount: '500'
        leafNodesToSearchPercent: 7
    approximateNeighborsCount: 150
    dimensions: 2
    distanceMeasureType: DOT_PRODUCT_DISTANCE
    shardSize: SHARD_SIZE_MEDIUM
metadataSchemaUri: gs://google-cloud-aiplatform/schema/matchingengine/metadata/nearest_neighbor_search_1.0.0.yaml
name: projects/415528280434/locations/us-west1/indexes/8192434750279909376
updateTime: '2022-11-13T16:35:40.571748Z'

➜  magic-modules git:(vertex-ai-index) ✗
gsutil cat gs://shotarok-vertex-ai-index-test/contents-v2/data.json
{"id": "1", "embedding": [0.1, 1.0], "restricts": [{"namespace": "class", "allow": ["cat", "pet"]},{"namespace": "category", "allow": ["feline"]}]}
{"id": "2", "embedding": [0.2, 1.0], "restricts": [{"namespace": "class", "allow": ["dog", "pet"]},{"namespace": "category", "allow": ["canine"]}]}
{"id": "3", "embedding": [0.3, 1.0], "restricts": [{"namespace": "class", "allow": ["cat", "pet"]},{"namespace": "category", "allow": ["feline"]}]}
{"id": "4", "embedding": [0.4, 1.0], "restricts": [{"namespace": "class", "allow": ["dog", "pet"]},{"namespace": "category", "allow": ["canine"]}]}

➜  magic-modules git:(vertex-ai-index) ✗
curl -X PATCH -H "Content-Type: application/json" -H "Authorization: Bearer $(gcloud auth print-access-token)" https://us-west1-aiplatform.googleapis.com/v1/projects/415528280434/locations/us-west1/indexes/8192434750279909376\?updateMask\=metadata.contentsDeltaUri,metadata.isCompleteOverwrite -d '{metadata: {contentsDeltaUri: "gs://shotarok-vertex-ai-index-test/contents-v2", isCompleteOverwrite: true}}'
{
  "name": "projects/415528280434/locations/us-west1/indexes/8192434750279909376/operations/6149817023779045376",
  "metadata": {
    "@type": "type.googleapis.com/google.cloud.aiplatform.v1.UpdateIndexOperationMetadata",
    "genericMetadata": {
      "createTime": "2022-11-13T17:38:18.050738Z",
      "updateTime": "2022-11-13T17:38:18.050738Z"
    }
  },
  "done": true,
  "response": {
    "@type": "type.googleapis.com/google.cloud.aiplatform.v1.Index",
    "name": "projects/415528280434/locations/us-west1/indexes/8192434750279909376"
  }
}

➜  magic-modules git:(vertex-ai-index) ✗
gcloud ai operations describe 6149817023779045376 --region us-west1 --index=8192434750279909376 --project=kouzoh-p-kohama
Using endpoint [https://us-west1-aiplatform.googleapis.com/]
done: true
metadata:
  '@type': type.googleapis.com/google.cloud.aiplatform.v1.UpdateIndexOperationMetadata
  genericMetadata:
    createTime: '2022-11-13T17:38:18.050738Z'
    updateTime: '2022-11-13T17:38:18.050738Z'
name: projects/415528280434/locations/us-west1/indexes/8192434750279909376/operations/6149817023779045376
response:
  '@type': type.googleapis.com/google.cloud.aiplatform.v1.Index
  name: projects/415528280434/locations/us-west1/indexes/8192434750279909376

➜  magic-modules git:(vertex-ai-index) ✗
gcloud ai indexes list --project=kouzoh-p-kohama --region=us-west1
Using endpoint [https://us-west1-aiplatform.googleapis.com/]
---
createTime: '2022-11-13T15:44:06.169597Z'
displayName: a test index (take 2)
etag: AMEw9yO5KRlZw_ZrKS6fWf_zZyWFe6RFbw5K0jNOC9WXDve__DOqFpYOtLiRStrxOL-e
indexStats:
  shardsCount: 1
  vectorsCount: '2'
indexUpdateMethod: BATCH_UPDATE
metadata:
  config:
    algorithmConfig:
      treeAhConfig:
        leafNodeEmbeddingCount: '500'
        leafNodesToSearchPercent: 7
    approximateNeighborsCount: 150
    dimensions: 2
    distanceMeasureType: DOT_PRODUCT_DISTANCE
    shardSize: SHARD_SIZE_MEDIUM
metadataSchemaUri: gs://google-cloud-aiplatform/schema/matchingengine/metadata/nearest_neighbor_search_1.0.0.yaml
name: projects/415528280434/locations/us-west1/indexes/8192434750279909376
updateTime: '2022-11-13T17:27:32.809234Z'
```

</details>

Reply:
I updated it in [d55968c](https://github.com/GoogleCloudPlatform/magic-modules/pull/6728/commits/d55968c7041b31f27b62691f9cd9a37acc8c59d8).